// email-templates/lib/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ensureDirectory = ensureDirectory;
exports.readContents = readContents;
exports.renderFile = renderFile;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _fs = require('fs');

var _glob = require('glob');

var _glob2 = _interopRequireDefault(_glob);

var _templateManager = require('./template-manager');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var readFileP = _bluebird2.default.promisify(_fs.readFile);
var globP = _bluebird2.default.promisify(_glob2.default);

function ensureDirectory(path, callback) {
  return new _bluebird2.default(function (resolve, reject) {
    (0, _fs.stat)(path, function (err, stat) {
      if (err) return reject(err);
      if (!stat.isDirectory()) return reject();
      resolve();
    });
  }).nodeify(callback);
}

function sailsViewsRender(path, type, cb){
  if (sails.hooks.views && sails.hooks.views.render) {
    var relPath = pathModule.relative(sails.config.paths.views, path + '/' + type);
    sails.hooks.views.render(relPath, {layout: '../email_layout'}, cb);
  } else {
    cb("sails hooks not available");
  }
}

function readContents(path, type, cb) {
  var filePath = path + '/*' + type + '.*';
  return globP(filePath).then(function (files) {
    if (!files.length) return null;

    return readFileP(files[0], 'utf8').then(function (content){
      if(!content){
        return null;
      }
      return {
        filename: files[0],
        content: content
      };
    });
  });
}

function renderFile(file, options) {
  if (!file) return _bluebird2.default.resolve(null);
  return (0, _templateManager.render)(file, options);
}
